File: interactive_engine/executor/Pegasus/src/operator/advanced/sink.rs
//
//! Copyright 2020 Alibaba Group Holding Limited.
//! 
//! Licensed under the Apache License, Version 2.0 (the "License");
//! you may not use this file except in compliance with the License.
//! You may obtain a copy of the License at
//! 
//!     http://www.apache.org/licenses/LICENSE-2.0
//! 
//! Unless required by applicable law or agreed to in writing, software
//! distributed under the License is distributed on an "AS IS" BASIS,
//! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//! See the License for the specific language governing permissions and
//! limitations under the License.

use super::*;
use crossbeam_channel::Sender;
use crate::operator::unary::Unary;
use crate::communication::Pipeline;

pub trait Output<D: Data, A> {
    fn sink<F>(&self, func: F) where F: Fn(u32, Vec<D>) + Send + Sync + 'static;

    fn sink_to<S: Sink<D> + Send + 'static>(&self, sink: S);
}

pub trait Sink<T: Data> {
    fn sink(&self, tag: u32, data: Vec<T>);
}

impl<T: Data> Sink<T> for Sender<(u32, Vec<T>)> {
    fn sink(&self, tag: u32, data: Vec<T>) {
        self.send((tag, data)).unwrap()
    }
}

impl<D: Data, A: DataflowBuilder> Output<D, A> for Stream<D, A> {

    fn sink<F>(&self, func: F) where F: Fn(u32, Vec<D>) + Send + Sync + 'static {
        self.unary("sink", Pipeline, |info| {
            info.set_sink();
            move |input, _out: &mut OutputHandle<()>| {
                input.for_each_batch(|data_set| {
                    let (t, d) = data_set.take();
                    let tag = t.current();
                    func(tag, d);
                    Ok(true)
                })?;
                Ok(())
            }
        });
    }

    fn sink_to<S: Sink<D> + Send + 'static>(&self, sink: S) {
        self.unary("sink", Pipeline, |info| {
            info.set_sink();
            move |input, _output: &mut OutputHandle<()>| {
                input.for_each_batch(|data_set| {
                    let (t, d) = data_set.take();
                    let tag = t.current();
                    sink.sink(tag, d);
                    Ok(true)
                })?;
                Ok(())
            }
        });
    }
}

File: flex/engines/graph_db/runtime/execute/ops/update/sink.cc
/** Copyright 2020 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "flex/engines/graph_db/runtime/execute/ops/update/sink.h"
namespace gs {
namespace runtime {
namespace ops {

class SinkInsertOpr : public IInsertOperator {
 public:
  SinkInsertOpr() {}

  std::string get_operator_name() const override { return "SinkInsertOpr"; }

  template <typename GraphInterface>
  bl::result<gs::runtime::WriteContext> eval_impl(
      GraphInterface& graph, const std::map<std::string, std::string>& params,
      gs::runtime::WriteContext&& ctx, gs::runtime::OprTimer& timer) {
    return ctx;
  }

  bl::result<gs::runtime::WriteContext> Eval(
      gs::runtime::GraphInsertInterface& graph,
      const std::map<std::string, std::string>& params,
      gs::runtime::WriteContext&& ctx, gs::runtime::OprTimer& timer) override {
    return eval_impl(graph, params, std::move(ctx), timer);
  }

  bl::result<gs::runtime::WriteContext> Eval(
      gs::runtime::GraphUpdateInterface& graph,
      const std::map<std::string, std::string>& params,
      gs::runtime::WriteContext&& ctx, gs::runtime::OprTimer& timer) override {
    return eval_impl(graph, params, std::move(ctx), timer);
  }
};

std::unique_ptr<IInsertOperator> SinkInsertOprBuilder::Build(
    const Schema& schema, const physical::PhysicalPlan& plan, int op_idx) {
  return std::make_unique<SinkInsertOpr>();
}

class USinkOpr : public IUpdateOperator {
 public:
  USinkOpr(const std::vector<int>& tag_ids) : tag_ids(tag_ids) {}

  std::string get_operator_name() const override { return "USinkOpr"; }

  bl::result<Context> Eval(GraphUpdateInterface& graph,
                           const std::map<std::string, std::string>& params,
                           Context&& ctx, OprTimer& timer) override {
    ctx.tag_ids = tag_ids;
    return ctx;
  }
  std::vector<int> tag_ids;
};

std::unique_ptr<IUpdateOperator> USinkOprBuilder::Build(
    const Schema& schema, const physical::PhysicalPlan& plan, int op_idx) {
  auto& opr = plan.plan(op_idx).opr().sink();
  std::vector<int> tag_ids;
  for (auto& tag : opr.tags()) {
    tag_ids.push_back(tag.tag().value());
  }
  return std::make_unique<USinkOpr>(tag_ids);
}

}  // namespace ops
}  // namespace runtime
}  // namespace gs